duckduckgo/BrowserServicesKit
# Feature flagging with autofill subfeatures

Please review the release process for BrowserServicesKit [here](https://app.asana.com/0/1200194497630846/1200837094583426).

**Required**:

Task/Issue URL: https://app.asana.com/0/72649045549333/1203676877763836
iOS PR:  https://github.com/duckduckgo/iOS/pull/1555
macOS PR: https://github.com/duckduckgo/macos-browser/pull/1044
What kind of version bump will this require?: Minor

**Optional**:

Tech Design URL:
https://app.asana.com/0/481882893211075/1203938569856702
https://app.asana.com/0/481882893211075/1203984214904257
CC: 

**Description**:

This PR:
- adds subfeatures support to the Privacy Config component
- moves the FeatureFlagger from iOS into BSK and updates it for the requirements gathered during this project
- moves the InternalUserDecider from iOS into BSK

To understand why I ended up at the solution I did, it's worth pointing out some requirements:
- We should be able to remotely toggle features using the privacy config
- Remote toggling shouldn't necessarily be possible while the feature is in development, only when we're close to release. Up until this point, we should stick with having them enabled internally alone by authenticating as an internal user.
- The current state of a given feature (particularly during development) may well be different in iOS vs macOS and we need a way to express that.
- Some features (autofill included) may be developed (and in the future disabled) only partially, introducing the need for nested features ("subfeatures") in the Privacy Configuration.



**Steps to test this PR**:
1. Check out the iOS & macOS PRs above
2. Check that any features configured using the privacy-config still work as expected (particularly autofill for iOS).
3. Check that any logic relying on the InternalUserDecider e.g:
    - internal updates and internal app icon on macOS
    - feature flags on iOS
4. Try changing the FeatureFlag configurations from iOS to make sure that they give the appropriate level of visibility.



**OS Testing**:

* [x] macOS 13

---
###### Internal references:
[Software Engineering Expectations](https://app.asana.com/0/59792373528535/199064865822552)
[Technical Design Template](https://app.asana.com/0/59792373528535/184709971311943)
